homeassistant/components/media_player/cast.py

```diff
@@ -33,92 +39,172 @@
+        enable_discovery = True
+    elif want_host[1] != DEFAULT_PORT:
+        # We're trying to add a group, so we have to use pychromecast's
+        # discovery to get the correct friendly name.
+        enable_discovery = True
+
+    if enable_discovery:
+        @callback
+        def async_cast_discovered(chromecast):
+            """Callback for when a new chromecast is discovered."""
+            if want_host is not None and \
+                    (chromecast.host, chromecast.port) != want_host:
+                return  # for groups, only add requested device
+            cast_device = _async_create_cast_device(hass, chromecast)
+
+            if cast_device is not None:
+                async_add_devices([cast_device])
+
+        async_dispatcher_connect(hass, SIGNAL_CAST_DISCOVERED,
+                                 async_cast_discovered)
+        # Re-play the callback for all past chromecasts, store the objects in
+        # a list to avoid concurrent modification resulting in exception.
+        for chromecast in list(hass.data[KNOWN_CHROMECASTS_KEY].values()):
+            async_cast_discovered(chromecast)
+
+        hass.async_add_job(_setup_internal_discovery, hass)
     else:
-        hosts = [tuple(dev[:2]) for dev in pychromecast.discover_chromecasts()
-                 if tuple(dev[:2]) not in known_hosts]
-
-    casts = []
-
-    # get_chromecasts() returns Chromecast objects with the correct friendly
-    # name for grouped devices
-    all_chromecasts = pychromecast.get_chromecasts()
-
-    for host in hosts:
-        (_, port) = host
-        found = [device for device in all_chromecasts
-                 if (device.host, device.port) == host]
-        if found:
-            try:
-                casts.append(CastDevice(found[0]))
-                known_hosts.append(host)
-            except pychromecast.ChromecastConnectionError:
-                pass
-
-        # do not add groups using pychromecast.Chromecast as it leads to names
-        # collision since pychromecast.Chromecast will get device name instead
-        # of group name
-        elif port == DEFAULT_PORT:
-            try:
-                # add the device anyway, get_chromecasts couldn't find it
-                casts.append(CastDevice(pychromecast.Chromecast(*host)))
-                known_hosts.append(host)
-            except pychromecast.ChromecastConnectionError:
-                pass
-
-    add_devices(casts)
+        # Manually add a "normal" Chromecast, we can do that without discovery.
+        try:
+            chromecast = pychromecast.Chromecast(*want_host)
+        except pychromecast.ChromecastConnectionError:
+            _LOGGER.warning("Can't set up chromecast on %s", want_host[0])
+            raise
+        key = (chromecast.host, chromecast.port, chromecast.uuid)
+        cast_device = _async_create_cast_device(hass, chromecast)
+        if cast_device is not None:
+            hass.data[KNOWN_CHROMECASTS_KEY][key] = chromecast
```

Comment:
Do you need to disconnect the chromecast object if you are not going to use it?

Reply:
I had that in there before, but then I remembered an edge case: what if `_async_create_cast_device` results in the last branch with setting the Chromecast object? Then we'd be invalidating the Chromecast object we just set. I left it out here since it's still much better than the previous code that just created every Chromecast object for every single `platform:` definition.

Reply:
And I think that could best be done with another refactor with `add_reader`.